evm/src/poseidon_sponge/poseidon_sponge_stark.rs

```diff
@@ -0,0 +1,657 @@
+use std::borrow::Borrow;
+use std::iter::once;
+use std::marker::PhantomData;
+
+use itertools::Itertools;
+use plonky2::field::extension::{Extendable, FieldExtension};
+use plonky2::field::packed::PackedField;
+use plonky2::field::polynomial::PolynomialValues;
+use plonky2::field::types::Field;
+use plonky2::hash::hash_types::RichField;
+use plonky2::hash::poseidon::Poseidon;
+use plonky2::iop::ext_target::ExtensionTarget;
+use plonky2::timed;
+use plonky2::util::timing::TimingTree;
+
+use super::columns::{
+    PoseidonSpongeColumnsView, NUM_POSEIDON_SPONGE_COLUMNS, POSEIDON_SPONGE_COL_MAP,
+};
+use crate::constraint_consumer::{ConstraintConsumer, RecursiveConstraintConsumer};
+use crate::cross_table_lookup::Column;
+use crate::evaluation_frame::{StarkEvaluationFrame, StarkFrame};
+use crate::poseidon::columns::{POSEIDON_SPONGE_RATE, POSEIDON_SPONGE_WIDTH};
+use crate::poseidon_sponge::columns::NUM_DIGEST_ELEMENTS;
+use crate::stark::Stark;
+use crate::util::trace_rows_to_poly_values;
+use crate::witness::memory::MemoryAddress;
+
+pub(crate) fn ctl_looked_data<F: Field>() -> Vec<Column<F>> {
+    let cols = POSEIDON_SPONGE_COL_MAP;
+    // We only take the first NUM_DIGEST_ELEMENTS, divided into two 32-bits limb elements, for the digest.
+    let outputs: Vec<_> =
+        Column::singles(&cols.output_rate_limbs[0..2 * NUM_DIGEST_ELEMENTS]).collect();
+    Column::singles([
+        cols.context,
+        cols.segment,
+        cols.virt,
+        cols.len,
+        cols.timestamp,
+    ])
+    .chain(outputs)
+    .collect()
+}
+
+pub(crate) fn ctl_looking<F: Field>() -> Vec<Column<F>> {
+    let cols = POSEIDON_SPONGE_COL_MAP;
+    let mut inputs: Vec<_> = (0..POSEIDON_SPONGE_RATE)
+        .map(|i| {
+            Column::linear_combination([
+                (cols.state_rate_limbs[2 * i], F::ONE),
+                (
+                    cols.state_rate_limbs[2 * i + 1],
+                    F::from_canonical_u64(1 << 32),
+                ),
+            ])
+        })
+        .collect();
+    inputs.extend(Column::singles(&cols.state_capacity));
```

Comment:
Similar concern as in PoseidonStark trace generation: if we allow for full 64-bit elements per Poseidon state cell, we may have either:
- undefined behavior if we call `from_canonical_u64`
- collisions if we call `from_noncanonical_u64` (as hashing `[p+1]` and `[1]` would yield the same digest).